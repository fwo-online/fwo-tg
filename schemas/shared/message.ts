import type { CharacterPublic } from '@/character/characterPublic';
import type { Character } from '@/character';
import type { Player } from './player';
import type { GameStatus } from '@/game';
import type { Action } from './action';
import type { RPC } from './rpc';

export type ClientToServerMessage = Message<{
  character: [callback: (character: Character) => void];
  'lobby:enter': [callback: (characters: CharacterPublic[]) => void];
  'lobby:leave': [];
  'lobby:start': [];
  'lobby:stop': [];
  'game:connected': [callback: (payload: RPC<{ players: Record<string, Player> }>) => void];
  'game:order': [
    order: {
      power: number;
      action: string;
      target: string;
    },
    callback: (
      payload: RPC<{
        actions: Action[];
        magics: Action[];
        skills: Action[];
        power: number;
        orders: {
          power: number;
          action: string;
          target: string;
        }[];
      }>,
    ) => void,
  ];
}>;

type Message<T extends Record<string, unknown[]>> = {
  [K in keyof T]: (...args: T[K]) => void;
};

export type ServerToClientMessage = Message<{
  character: [character: Character];
  'lobby:enter': [character: CharacterPublic];
  'lobby:leave': [character: CharacterPublic];
  'lobby:list': [characters: CharacterPublic[]];
  'lobby:start': [character: CharacterPublic];
  'lobby:stop': [character: CharacterPublic];
  'game:start': [gameID: string];
  'game:end': [];
  'game:startOrders': [actions: { actions: Action[]; magics: Action[]; skills: Action[] }];
  'game:endOrders': [];
  'game:startRound': [
    {
      round: number;
      status: Partial<Record<string, GameStatus[]>>;
    },
  ];
  'game:endRound': [{ dead: Player[] }];
  'game:kick': [{ reason: string; player: Player }];
  'game:preKick': [{ reason: string; player: Player }];
}>;
