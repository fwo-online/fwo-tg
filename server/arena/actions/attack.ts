import type { Affect } from '../Constuructors/interfaces/Affect';
import PhysConstructor from '../Constuructors/PhysConstructor';
import CastError from '../errors/CastError';
/**
 * Физическая атака
 */
class Attack extends PhysConstructor implements Affect {
  constructor() {
    super({
      name: 'attack',
      displayName: 'Атака',
      desc: 'Атака',
      lvl: 0,
      orderType: 'all',
    });
  }

  /**
 * Кастомный обработчик атаки
 */
  run() {
    if (!this.params) {
      return;
    }
    const { initiator, target } = this.params;

    target.flags.isHited = {
      initiator: initiator.nick, val: this.status.effect,
    };

    target.stats.down('hp', this.status.effect);
  }

  preAffect: Affect['preAffect'] = ({ params: { initiator, target, game } }): undefined => {
    if (target.flags.isHited) {
      throw new CastError(this.getSuccessResult({ initiator: target, target: initiator, game }));
    }
  };
}
export default new Attack();
