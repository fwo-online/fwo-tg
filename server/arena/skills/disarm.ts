import type { Affect } from '../Constuructors/interfaces/Affect';
import { Skill } from '../Constuructors/SkillConstructor';
import CastError from '../errors/CastError';

/**
 * Обезоруживание
 */
class Disarm extends Skill implements Affect {
  constructor() {
    super({
      name: 'disarm',
      displayName: '🥊 Обезоруживание',
      desc: 'Обезоруживает противника, не давая ему совершить атаку оружием',
      cost: [12, 13, 14, 15, 16, 17],
      proc: 10,
      baseExp: 20,
      costType: 'en',
      orderType: 'enemy',
      aoeType: 'target',
      chance: [70, 75, 80, 85, 90, 95],
      effect: [1.1, 1.2, 1.3, 1.4, 1.5, 1.6],
      profList: { w: 3, l: 5 },
      bonusCost: [10, 20, 30, 40, 60, 80],
    });
  }

  run() {
    const { initiator, target } = this.params;
    const initiatorMagicLvl = initiator.skills[this.name];
    const effect = this.effect[initiatorMagicLvl - 1] || 1;
    // изменяем
    const iDex = initiator.stats.val('attributes.dex') * effect;
    const tDex = target.stats.val('attributes.dex');
    if (iDex >= tDex) {
      target.flags.isDisarmed = true;

      this.calculateExp();
    } else {
      throw new CastError('SKILL_FAIL');
    }
  }

  preAffect: Affect['preAffect'] = ({ params: { initiator, target, game } }): undefined => {
    if (initiator.flags.isDisarmed) {
      throw new CastError(this.getSuccessResult({ initiator: target, target: initiator, game }));
    }
  };
}

export default new Disarm();
