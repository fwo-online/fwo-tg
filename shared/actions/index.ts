import type { CharacterClass } from '@/character';

export type Magic = {
  name: string;
  description: string;
  displayName: string;
  effect: string[];
  cost: number;
  costType: CostType;
  lvl: number;
  orderType: OrderType;
  effectType?: EffectType;
};

export type Skill = {
  name: string;
  displayName: string;
  description: string;
  classList: Partial<Record<CharacterClass, number>>;
  bonusCost: number[];
  orderType: OrderType;
  cost: number;
  costType: CostType;
};

export enum CostType {
  MP = 'mp',
  EN = 'en',
}

export enum OrderType {
  All = 'all',
  Any = 'any',
  Enemy = 'enemy',
  Self = 'self',
  Team = 'team',
  TeamExceptSelf = 'teamExceptSelf',
}

export enum EffectType {
  Acid = 'acid',
  Fire = 'fire',
  Lighting = 'lighting',
  Frost = 'frost',
  Physical = 'physical',
  Clear = 'clear',
}

export type ActionType =
  | 'magic'
  | 'dmg-magic'
  | 'dmg-magic-long'
  | 'aoe-dmg-magic'
  | 'magic-long'
  | 'skill'
  | 'phys'
  | 'heal-magic'
  | 'heal'
  | 'protect'
  | 'dodge'
  | 'passive'
  | 'regeneration';

export type ReservedFailReason =
  | 'NO_INITIATOR'
  | 'NO_TARGET'
  | 'NO_MANA'
  | 'NO_ENERGY'
  | 'CHANCE_FAIL'
  | 'GOD_FAIL'
  | 'HEAL_FAIL'
  | 'SKILL_FAIL'
  | 'PHYS_FAIL'
  | 'NO_WEAPON';
