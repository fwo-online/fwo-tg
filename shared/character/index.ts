import type { CharacterAttributes } from './characterAttributesSchema';
import { CharacterClass } from './characterClassSchema';
import type { Attributes } from '@/shared/attributes';

export const characterAttributesByClass: Record<CharacterClass, CharacterAttributes> = {
  [CharacterClass.Warrior]: {
    str: 10,
    dex: 8,
    int: 3,
    wis: 3,
    con: 6,
  },
  [CharacterClass.Archer]: {
    str: 3,
    dex: 10,
    int: 8,
    wis: 3,
    con: 6,
  },
  [CharacterClass.Mage]: {
    str: 3,
    dex: 3,
    int: 8,
    wis: 10,
    con: 6,
  },
  [CharacterClass.Priest]: {
    str: 3,
    dex: 3,
    int: 10,
    wis: 8,
    con: 6,
  },
};

export type CharacterDynamicAttributes = Attributes & {
  maxTarget: number;
  spellLength: number;
};

export * from './characterAttributesSchema';
export * from './characterClassSchema';
export * from './characterSchema';
export * from './createCharacterSchema';
export * from './characterPublic';
