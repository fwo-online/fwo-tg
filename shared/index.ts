export * from './character';
export * from './inventory';
export * from './item';
export * from './itemSet';
export * from './shared';
export * from './actions';
export * from './game';
export * from './clan';
