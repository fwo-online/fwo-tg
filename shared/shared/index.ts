export * from './id';
export * from './modifiers';
export * from './player';
export * from './orderSchema';
export * from './message';
export * from './attributes';
export * from './action';
