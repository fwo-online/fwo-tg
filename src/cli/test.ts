import { $ } from 'bun';

async function test() {
  try {
    await $`docker-compose up -d --build db`;

    await $`bun test --preload ./test.setup.ts`;

    await $`docker-compose stop db`;
    process.exit(0);
  } catch {
    process.exit(1);
  }
}

void test();
