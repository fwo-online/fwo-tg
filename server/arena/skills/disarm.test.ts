import { afterEach, beforeEach, describe, expect, it } from 'bun:test';
import { CharacterClass } from '@fwo/shared';
import type GameService from '@/arena/GameService';
import TestUtils from '@/utils/testUtils';
import attack from '../actions/attack';
import disarm from './disarm';

// npm t server/arena/skills/disarm.test.ts

describe('disarm', () => {
  let game: GameService;

  beforeEach(async () => {
    attack.registerPreAffects([disarm]);

    game = await TestUtils.createGame([
      {
        prof: CharacterClass.Warrior,
        weapon: {},
      },
      {
        prof: CharacterClass.Archer,
        skills: { disarm: 3 },
      },
      {
        prof: CharacterClass.Archer,
        skills: { disarm: 3 },
      },
    ]);

    TestUtils.mockRandom();
  });

  afterEach(() => {
    TestUtils.restoreRandom();
  });

  it('target should be disarmed if initiator has more dex', async () => {
    game.players.players[0].proc = 1;
    game.players.players[1].stats.set('attributes.dex', 9999);

    disarm.cast(game.players.players[1], game.players.players[0], game);
    attack.cast(game.players.players[0], game.players.players[1], game);

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('target should not be disarmed if initiator has less dex', async () => {
    game.players.players[0].proc = 1;
    game.players.players[0].stats.set('attributes.dex', 9999);

    disarm.cast(game.players.players[1], game.players.players[0], game);
    attack.cast(game.players.players[0], game.players.players[1], game);

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('should handle several casters', async () => {
    game.players.players[0].proc = 1;
    game.players.players[1].stats.set('attributes.dex', 9999);
    game.players.players[2].stats.set('attributes.dex', 9999);

    disarm.cast(game.players.players[1], game.players.players[0], game);
    disarm.cast(game.players.players[2], game.players.players[0], game);
    attack.cast(game.players.players[0], game.players.players[1], game);

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
