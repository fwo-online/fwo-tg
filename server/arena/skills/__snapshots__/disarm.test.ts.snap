// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`disarm target should be disarmed if initiator has more dex 1`] = `
"*alias* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _ü•ä –û–±–µ–∑–æ—Ä—É–∂–∏–≤–∞–Ω–∏–µ_ –Ω–∞ *asperiores*
\\[ üìñ20 ]
*asperiores* –ø—ã—Ç–∞–ª—Å—è –∞—Ç–∞–∫–æ–≤–∞—Ç—å *alias*, –Ω–æ —É –Ω–µ–≥–æ –Ω–µ –ø–æ–ª—É—á–∏–ª–æ—Å—å
_ü•ä –û–±–µ–∑–æ—Ä—É–∂–∏–≤–∞–Ω–∏–µ_ *alias*: üìñ0"
`;

exports[`disarm target should not be disarmed if initiator has less dex 1`] = `
"*explicabo* –ø—ã—Ç–∞–ª—Å—è –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å —É–º–µ–Ω–∏–µ _ü•ä –û–±–µ–∑–æ—Ä—É–∂–∏–≤–∞–Ω–∏–µ_, –Ω–æ —É –Ω–µ–≥–æ –Ω–µ –≤—ã—à–ª–æ
*repellat* —Ä—É–±–∞–Ω—É–ª *explicabo* _–≠–ø–∏—á–µ—Å–∫–∏–º –ú–µ—á–æ–º_ –∏ –Ω–∞–Ω—ë—Å *4.85* —É—Ä–æ–Ω–∞
\\[ explicabo üí• üíî-4.85/3.15 üìñ39 ]"
`;
